tests/utils/managed_utils/mod.rs

```diff
@@ -0,0 +1,583 @@
+            // For now, everyone supports all ciphersuites.
+            let mut credential_bundles = Vec::new();
+            for ciphersuite in &Config::supported_ciphersuite_names() {
+                let credential_bundle =
+                    CredentialBundle::new(identity.clone(), CredentialType::Basic, *ciphersuite)
+                        .unwrap();
+                credential_bundles.push((*ciphersuite, credential_bundle));
+            }
+            key_store.store_credentials(identity.clone(), credential_bundles);
+        }
+        let clients = RefCell::new(HashMap::new());
+        let groups = RefCell::new(HashMap::new());
+        let waiting_for_welcome = RefCell::new(HashMap::new());
+        ManagedTestSetup {
+            number_of_clients,
+            clients,
+            groups,
+            key_store,
+            waiting_for_welcome,
+            default_mgc,
+        }
+    }
+
+    pub fn create_clients(&'ks self) {
+        for i in 0..self.number_of_clients {
+            let identity = i.to_be_bytes().to_vec();
+            // For now, everyone supports all ciphersuites.
+            let _ciphersuites = Config::supported_ciphersuite_names();
+            let mut credential_bundles = Vec::new();
+            let key_package_bundles = RefCell::new(HashMap::new());
+            let client = Client {
+                identity: identity.clone(),
+                _ciphersuites,
+                key_store: &self.key_store,
+                key_package_bundles,
+                groups: RefCell::new(HashMap::new()),
+            };
+            for ciphersuite in &Config::supported_ciphersuite_names() {
+                let credential_bundle =
+                    CredentialBundle::new(identity.clone(), CredentialType::Basic, *ciphersuite)
+                        .unwrap();
+                credential_bundles.push((*ciphersuite, credential_bundle));
+            }
+            let mut clients = self.clients.borrow_mut();
+            clients.insert(identity, RefCell::new(client));
+            drop(clients);
+        }
+    }
+
+    /// Create a fresh `KeyPackage` for client `client` for use when adding it
+    /// to a group. The `KeyPackageBundle` will be fetched automatically when
+    /// delivering the `Welcome` via `deliver_welcome`.
+    pub fn get_fresh_key_package(
+        &self,
+        client: &Client,
+        ciphersuite: &Ciphersuite,
+    ) -> Result<KeyPackage, SetupError> {
+        let key_package = client.get_fresh_key_package(ciphersuite)?;
+        println!("Storing key package with hash: {:?}", key_package.hash());
+        self.waiting_for_welcome
+            .borrow_mut()
+            .insert(key_package.hash(), client.identity.clone());
+        Ok(key_package)
+    }
+
+    /// Deliver a Welcome message to group `group` to the intended recipients.
+    pub fn deliver_welcome(&self, welcome: Welcome, group: &Group) -> Result<(), SetupError> {
```

Comment:
You pass in the `group` here and pull information from it. If that's intended in you test setup, it needs description of how it is supposed to be used and how it allows testing things. In the real world the `group_config` must be extracted from the `Welcome` and the `public_tree` might be in there as well, or delivered separately. (All things that are impossible to test with this function.)

Reply:
Fair point. I guess I wrote this function mostly with `pub(crate)` use in mind, but it's certainly something we will want to use outside. I should probably switch from passing in `&Group` everywhere to passing in `GroupId`.

Reply:
Made the documentation more precise.